lib/resourcemerge/apiext.go

```diff
@@ -12,18 +12,36 @@ import (
 // modified is set to true when existing had to be updated with required.
 func EnsureCustomResourceDefinitionV1(modified *bool, existing *apiextv1.CustomResourceDefinition, required apiextv1.CustomResourceDefinition) {
 	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+	ensureCustomResourceDefinitionV1Defaults(&required)
+	ensureCustomResourceDefinitionV1CaBundle(&required, *existing)
+	if !equality.Semantic.DeepEqual(existing.Spec, required.Spec) {
+		*modified = true
+		existing.Spec = required.Spec
+	}
+}
 
-	// apply defaults to blue print
+func ensureCustomResourceDefinitionV1Defaults(required *apiextv1.CustomResourceDefinition) {
 	if len(required.Spec.Names.Singular) == 0 {
 		required.Spec.Names.Singular = strings.ToLower(required.Spec.Names.Kind)
 	}
 	if len(required.Spec.Names.ListKind) == 0 {
 		required.Spec.Names.ListKind = required.Spec.Names.Kind + "List"
 	}
+}
 
-	// we stomp everything
-	if !equality.Semantic.DeepEqual(existing.Spec, required.Spec) {
-		*modified = true
-		existing.Spec = required.Spec
+// ensureCustomResourceDefinitionV1CaBundle ensures that the field
+// .Spec.Conversion.Webhook.ClientConfig.CABundle is not managed by the CVO when
+// the service-ca controller is responsible for the field.
+func ensureCustomResourceDefinitionV1CaBundle(required *apiextv1.CustomResourceDefinition, existing apiextv1.CustomResourceDefinition) {
+	if val, ok := existing.ObjectMeta.Annotations[injectCABundleAnnotation]; !ok || val != "true" {
+		return
+	}
+	if required.Spec.Conversion != nil && required.Spec.Conversion.Strategy == apiextv1.WebhookConverter {
+		// The service CA bundle is only injected by the service-ca controller into
+		// the CRD if the CRD is configured to use a webhook for conversion
+		if required.Spec.Conversion.Webhook != nil && required.Spec.Conversion.Webhook.ClientConfig != nil &&
+			existing.Spec.Conversion != nil && existing.Spec.Conversion.Webhook != nil && existing.Spec.Conversion.Webhook.ClientConfig != nil {
```

Comment:
Putting all 5 conditions in one if conditional statement is a good practice. This reduces the readability of the code. Is there anyway we can break it down to nested if statements or someway make the code simpler?

Reply:
Do we need to really check `existing.Spec.Conversion != nil` and   `existing.Spec.Conversion.Webhook` if we are checking for `existing.Spec.Conversion.Webhook.ClientConfig != nil`.

Reply:
> Is there anyway we can break it down to nested if statements or someway make the code simpler?

Yes, we can definitely try. I have restructured the conditions and added an `if` statement as to differentiate the checks for `required` and `existing`. What do you think about something like this?

```go
if required.Spec.Conversion != nil && required.Spec.Conversion.Strategy == apiextv1.WebhookConverter {
	// The service CA bundle is only injected by the service-ca controller into
	// the CRD if the CRD is configured to use a webhook for conversion
	if required.Spec.Conversion.Webhook != nil &&
		required.Spec.Conversion.Webhook.ClientConfig != nil {
		if existing.Spec.Conversion != nil &&
			existing.Spec.Conversion.Webhook != nil &&
			existing.Spec.Conversion.Webhook.ClientConfig != nil {
			required.Spec.Conversion.Webhook.ClientConfig.CABundle = existing.Spec.Conversion.Webhook.ClientConfig.CABundle
		}
	}
}
```

---

> Do we need to really check existing.Spec.Conversion != nil and existing.Spec.Conversion.Webhook if we are checking for existing.Spec.Conversion.Webhook.ClientConfig != nil.

Yes, I do believe so. All these fields are pointers and need to be checked before dereferencing, so we don't dereference a `nil` by any chance.

For example, in the [openshift/library-go](https://github.com/openshift/library-go) we can find an example of accessing the `<apiextensionsv1.CustomResourceDefinition>.Spec.Conversion.Webhook.ClientConfig` field like this ([pkg/operator/resource/resourcemerge/apiextensions.go#L41-L46](https://github.com/openshift/library-go/blob/e555322cb70844f90b003cbdfe7ce002d7c61810/pkg/operator/resource/resourcemerge/apiextensions.go#L41-L46)):

```go
func mimicCRDV1Defaulting(required *apiextensionsv1.CustomResourceDefinition) {
	crd_SetDefaults_CustomResourceDefinitionSpec(&required.Spec)

	if required.Spec.Conversion != nil &&
		required.Spec.Conversion.Webhook != nil &&
		required.Spec.Conversion.Webhook.ClientConfig != nil &&
		required.Spec.Conversion.Webhook.ClientConfig.Service != nil {
		crd_SetDefaults_ServiceReference(required.Spec.Conversion.Webhook.ClientConfig.Service)
	}
}
```

A similar thing at [openshift/operator-framework-olm/staging/operator-lifecycle-manager/pkg/controller/operators/olm/apiservices.go#L298-L302](https://github.com/openshift/operator-framework-olm/blob/55503f8eaa40ec8aa02d0b1310609c7551988554/staging/operator-lifecycle-manager/pkg/controller/operators/olm/apiservices.go#L298-L302) as well:

```go
if crd.Spec.Conversion == nil || crd.Spec.Conversion.Webhook == nil || crd.Spec.Conversion.Webhook.ClientConfig == nil && crd.Spec.Conversion.Webhook.ClientConfig.CABundle == nil {
	continue
}

return crd.Spec.Conversion.Webhook.ClientConfig.CABundle, nil
```

Reply:
![image](https://user-images.githubusercontent.com/3125299/209062748-a7f48754-6e13-4a84-91e3-aca467ab8a81.png)
This format looks ok to me. Thanks